### ISSUE TITLE
feat: Dropout layer for regularization

### ISSUE THREAD

**Issue by dndungu** [zerfoo/zerfoo#97]
## Summary

Training tabular models on financial data is highly prone to overfitting — the signal-to-noise ratio is very low (~50% directional accuracy baseline). Standard dropout between dense layers is one of the most effective regularization techniques for MLPs on tabular data.

Zerfoo has a feature dropout issue (#50) for dropping entire input features, but standard **element-wise dropout** on hidden layers is not yet available.

## Proposed API

```go
// In package layers/regularization or layers/core

// Dropout randomly zeros elements with probability p during training.
// During inference (when training=false), acts as identity.
type Dropout[T numeric.Float] struct {
    P        float64 // drop probability (e.g., 0.3)
    Training bool    // toggle for train vs eval mode
}

func NewDropout[T numeric.Float](engine compute.Engine[T], p float64) *Dropout[T]

// Graph-level train/eval toggle
func (g *Graph[T]) Train()  // enables dropout
func (g *Graph[T]) Eval()   // disables dropout
```

## Use Case

Wolf's MLP for COIN price prediction has 81-633 input features and 2 hidden layers. Without dropout, the model memorizes training noise (all experiments show ~50% OOS accuracy despite low training loss). Standard practice is:

```
Input → Dense(81→162) → ReLU → Dropout(0.3) → Dense(162→81) → ReLU → Dropout(0.3) → Dense(81→1) → Sigmoid
```

## Related

- #50 (feature dropout — complementary, not the same)
- #47 (early stopping — related regularization)

**Comment by dndungu** [comment]
Implemented in `layers/regularization/dropout.go`. Closing as resolved.
